frame/staking/src/tests.rs

```diff
@@ -4129,57 +4125,43 @@ mod election_data_provider {
 	#[test]
 	fn get_max_len_voters_even_if_some_nominators_are_slashed() {
 		ExtBuilder::default()
-			.nominate(true) // add nominator 101, who nominates [11, 21]
+			.nominate(false)
 			.add_staker(61, 60, 20, StakerStatus::<AccountId>::Nominator(vec![21]))
-			//                                 61 only nominates validator 21 ^^
 			.add_staker(71, 70, 10, StakerStatus::<AccountId>::Nominator(vec![11, 21]))
+			.add_staker(81, 80, 10, StakerStatus::<AccountId>::Nominator(vec![11, 21]))
 			.build_and_execute(|| {
-				// given our nominators ordered by stake,
+				// given our voters ordered by stake,
 				assert_eq!(
 					<Test as Config>::SortedListProvider::iter().collect::<Vec<_>>(),
-					vec![101, 61, 71]
-				);
-
-				// and total voters
-				assert_eq!(
-					<Test as Config>::SortedListProvider::count() +
-						<Validators<Test>>::iter().count() as u32,
-					6
+					vec![11, 21, 31, 61, 71, 81]
 				);
 
-				// we take 5 voters
+				// we take 4 voters
 				assert_eq!(
-					Staking::voters(Some(5))
+					Staking::voters(Some(4))
 						.unwrap()
 						.iter()
 						.map(|(stash, _, _)| stash)
 						.copied()
 						.collect::<Vec<_>>(),
-					// then
-					vec![
-						31, 21, 11, // 3 nominators
-						101, 61 // 2 validators, and 71 is excluded
-					],
+					vec![11, 21, 31, 61],
 				);
 
 				// roll to session 5
 				run_to_block(25);
 
-				// slash 21, the only validator nominated by 61
+				// slash 21, the only validator nominated by 61 and 21.
 				add_slash(&21);
 
-				// we take 4 voters
+				// we take 4 voters; 71 and 81 are replacing the ejected ones.
 				assert_eq!(
 					Staking::voters(Some(4))
 						.unwrap()
 						.iter()
 						.map(|(stash, _, _)| stash)
 						.copied()
 						.collect::<Vec<_>>(),
-					vec![
-						31, 11, // 2 validators (21 was slashed)
-						101, 71 // 2 nominators, excluding 61
-					],
+					vec![11, 31, 71, 81],
```

Comment:
w.r.t to validator 21 getting removed, where in the code is a partially slashed validator would be removed from voters?

Reply:
When 21 is slashed, its vote weight is not updated in the bags list - so we should still iterate it second because the bags list has its pre-slash vote weight.

Reply:
when they receive a non-zero slash, some field in their slashing record is updated, which is then filtered out here: https://github.com/paritytech/substrate/blob/8ba2563f9d71c54289f9bfcf99dd0331118fb3bd/frame/staking/src/pallet/impls.rs#L689

Reply:
this link 404s 😅

Reply:
The only place I see slashing spans refered to is when filtering the nominators targets